Enable more clazy checks by default

Holzhaus opened the pull request.

Holzhaus committed:
fix(clazy): Declare `Q_PRIMITIVE_TYPE`/`Q_MOVABLE_TYPE` for types in lists

Holzhaus committed:
fix(clazy): Fix `-Wclazy-returning-void-expression` warnings

Holzhaus committed:
fix(clazy): Fix some `-Wclazy-missing-qobject-cast` warnings

Holzhaus committed:
fix(clazy): Fix or silence `-Wclazy-base-class-event` warnings

Holzhaus committed:
chore(clazy): Silence `-Wclazy-wrong-qevent-cast` warnings

Holzhaus committed:
ci(github): Enable more clazy checks in `build-checks` workflow

This enables the following checks:

- wrong-qevent-cast
- missing-typeinfo
- base-class-event

Swiftb0y commented on src/test/replaygaintest.cpp:
```diff
@@ -5,6 +5,10 @@
 
+// The gtest `ASSERT_*` macros return a void expression and thus may lead to a
+// lot of clazy warnings.
+// clazy:excludeall=returning-void-expression,missing-qobject-macro
```
why `missing-qobject-macro`?

Holzhaus replied:
We have some test classes that inherit from it. But adding Q_OBJECT won't help, because gtest generates new classes from it via macro magic, and we can't add Q_OBJECT to these anyway.

Swiftb0y replied:
makes sense, this should also go in the comment imo.

Swiftb0y commented on src/mixxxapplication.cpp:
```diff
@@ -154,7 +155,8 @@ bool MixxxApplication::notify(QObject* target, QEvent* event) {
     }
     case QEvent::MouseButtonRelease: {
-        QMouseEventEditable* mouseEvent = static_cast<QMouseEventEditable*>(event);
+        QMouseEventEditable* mouseEvent =
+                static_cast<QMouseEventEditable*>(event); // clazy:exclude=wrong-qevent-cast
```
I just looked at this recently, this actually triggers UBSAN, so we should find an alternative to `QMouseEventEditable` and this cast.

Holzhaus replied:
I agree. Fortunately this file seems to be the only occurrence so we should enable that clazy rule to avoid adding it elsewhere.

daschuer replied:
Qt seems to have a similar issue here: https://bugreports.qt.io/browse/QTBUG-99563

daschuer replied:
Normally a downcast to a inherited object is undefined, because the object might be bigger and we might access memory out of the boundary of the original object. I think the only way around the undefined behavior is to make a copy, modify the value and than copy the result back. That is a CPU cost to fix an non-issue   ... :-/

Swiftb0y replied:
I don't think these are particularly expensive... I'd take the hit for avoiding UB. It would be very annoying to use ubsan with "false positives" like this.

Holzhaus replied:
I opened an issue for this: https://github.com/mixxxdj/mixxx/issues/12013

Swiftb0y commented on .github/workflows/build-checks.yml:
```diff
@@ -117,7 +117,7 @@ jobs:
         working-directory: build
         env:
-          CLAZY_CHECKS: level2,no-rule-of-two-soft,no-non-pod-global-static,no-qproperty-without-notify,no-wrong-qevent-cast,no-qstring-allocations,no-function-args-by-value,no-copyable-polymorphic,no-ctor-missing-parent-argument,no-missing-qobject-macro,no-rule-of-three,no-returning-void-expression,no-missing-typeinfo,no-base-class-event
+          CLAZY_CHECKS: level2,no-rule-of-two-soft,no-non-pod-global-static,no-qproperty-without-notify,no-qstring-allocations,no-function-args-by-value,no-copyable-polymorphic,no-ctor-missing-parent-argument,no-rule-of-three
```
I don't think we should use `no-returning-void-expression`. It results in too much noise in the tests and the bug it protects against (at least in their example) is very unlikely to occur in mixxx due to our code formatting rules.

Swiftb0y replied:
I didn't spot and true positives for `no-returning-void-expression` in the diff. Did you?

daschuer replied:
https://github.com/mixxxdj/mixxx/blob/main/src/engine/controls/enginecontrol.cpp#L74

Swiftb0y replied:
unnecessary, sure, but not a bug.

Swiftb0y reviewed (changes requested):
Thank you, but this needs a little more work and discussion. IMO all `clazy:exclude`s need an explanation why its valid.

Holzhaus commented on src/engine/filters/enginefilteriir.h:
```diff
@@ -25,13 +25,11 @@ enum IIRPass {
 
-
-class EngineFilterIIRBase : public EngineObjectConstIn {
+class EngineFilterIIRBase : public EngineObjectConstIn { // clazy:exclude=missing-qobject-macro
```
I'm questioning the purpose of this class. We could just merge it into `EngineFilterIIR`:

```
$ git grep EngineFilterIIRBase
src/analyzer/analyzerwaveform.h:class EngineFilterIIRBase;
src/analyzer/analyzerwaveform.h:    EngineFilterIIRBase* m_filter[FilterCount];
src/engine/filters/enginefilteriir.h:class EngineFilterIIRBase : public EngineObjectConstIn {
src/engine/filters/enginefilteriir.h:class EngineFilterIIR : public EngineFilterIIRBase {
```

Holzhaus replied:
Ah, it's because EngineFilterIIR is a template class.

Holzhaus replied:
Anyway, we can't use it here because there is no CPP file for it and leads to linker errors:

```
mold: error: undefined symbol: EngineFilterIIRBase::qt_metacast(char const*)
>>> referenced by enginefilterlinkwitzriley4.cpp
>>>               libmixxx-lib.a(enginefilterlinkwitzriley4.cpp.o):(EngineFilterLinkwitzRiley4Low::qt_metacast(char const*))>>> referenced by enginefilterbessel4.cpp
>>>               libmixxx-lib.a(enginefilterbessel4.cpp.o):(EngineFilterBessel4Low::qt_metacast(char const*))>>> referenced by enginefilterbessel4.cpp
>>>               libmixxx-lib.a(enginefilterbessel4.cpp.o):(EngineFilterBessel4Band::qt_metacast(char const*))>>> referenced 13 more times

mold: error: undefined symbol: typeinfo for EngineFilterIIRBase
>>> referenced by enginefilterbessel4.cpp
>>>               libmixxx-lib.a(enginefilterbessel4.cpp.o)>>> referenced by enginefilterbessel4.cpp
>>>               libmixxx-lib.a(enginefilterbessel4.cpp.o)>>> referenced by enginefilterbessel4.cpp
>>>               libmixxx-lib.a(enginefilterbessel4.cpp.o)>>> referenced 7 more times

mold: error: undefined symbol: EngineFilterIIRBase::staticMetaObject
>>> referenced by enginefilterlinkwitzriley4.cpp
>>>               libmixxx-lib.a(enginefilterlinkwitzriley4.cpp.o)>>> referenced by enginefilterlinkwitzriley4.cpp
>>>               libmixxx-lib.a(enginefilterlinkwitzriley4.cpp.o)>>> referenced by enginefilterbessel4.cpp
>>>               libmixxx-lib.a(enginefilterbessel4.cpp.o)>>> referenced 13 more times

mold: error: undefined symbol: EngineFilterIIRBase::qt_metacall(QMetaObject::Call, int, void**)
>>> referenced by enginefilterbessel4.cpp
>>>               libmixxx-lib.a(enginefilterbessel4.cpp.o):(EngineFilterBessel4Low::qt_metacall(QMetaObject::Call, int, void**))>>> referenced by enginefilterlinkwitzriley4.cpp
>>>               libmixxx-lib.a(enginefilterlinkwitzriley4.cpp.o):(EngineFilterLinkwitzRiley4Low::qt_metacall(QMetaObject::Call, int, void**))>>> referenced by enginefilterbessel4.cpp
>>>               libmixxx-lib.a(enginefilterbessel4.cpp.o):(EngineFilterBessel4Band::qt_metacall(QMetaObject::Call, int, void**))>>> referenced 13 more times

clang-16: error: linker command failed with exit code 1 (use -v to see invocation)
```

Swiftb0y replied:
good point, QObjects can't be templated...

Holzhaus commented on src/library/multilineeditdelegate.cpp:
```diff
@@ -67,7 +67,7 @@ bool MultiLineEditor::eventFilter(QObject* obj, QEvent* event) {
                 ke->modifiers().testFlag(Qt::NoModifier)) {
             emit editingFinished();
-            return false;
+            return false; // clazy:exclude=base-class-event
```
I wasn't sure, it emits `editingFinished()` before. I can remove the `return false` but I was unsure if there are any sideeffects.

daschuer replied:
Just checked the Qt source and it should not make a difference. The base class returns false as well. 
So I would prefer to remove return false.

Holzhaus replied:
https://github.com/mixxxdj/mixxx/pull/12008/commits/93ccf8279586b49df807d57b1f7135cd634123ac

Holzhaus committed:
fix(wwidgetgroup): Fix `-Wclazy-qproperty-type-mismatch` warning

Fixes the following warning:

    src/widget/wwidgetgroup.h:63:5: error: Q_PROPERTY 'layoutAlignment' of type 'Qt::Alignment' is mismatched with method 'setLayoutAlignment' with parameter of type 'int' [-Wclazy-qproperty-type-mismatch]
        void setLayoutAlignment(int alignment);
        ^

Holzhaus committed:
fix(controlpickermenu): Reduce the number of `QList` allocations

Fixes the following warning:

    src/controllers/controlpickermenu.cpp:130:24: error: container inside loop causes unneeded allocations [-Wclazy-container-inside-loop]
            QList<QString> eqNames;
                           ^

Holzhaus committed:
fix(effectsbackend): Mark backend type names as translatable

Silences the following warnings:

    src/effects/backends/effectsbackend.cpp:33:16: error: tr() without a literal string [-Wclazy-tr-non-literal]
            return QObject::tr(backendTypeNameBuiltIn);
                   ^
    src/effects/backends/effectsbackend.cpp:37:16: error: tr() without a literal string [-Wclazy-tr-non-literal]
            return QObject::tr(backendTypeNameUnknown);
                   ^

Holzhaus committed:
fix: Use `!isEmpty()` instead of using `count()` as bool

Fixes warnings like these:

    src/engine/enginemixer.cpp:509:47: error: use isEmpty() instead [-Wclazy-isempty-vs-count]
            m_pTalkoverDucking->setAboveThreshold(m_activeTalkoverChannels.size());
                                                  ^

Holzhaus committed:
fix(browsetablemodel): Reserve row data beforehand and not inside loop

Fixes the following warning:

    src/library/browse/browsethread.cpp:154:31: error: container inside loop causes unneeded allocations [-Wclazy-container-inside-loop]
            QList<QStandardItem*> row_data;
                                  ^

Holzhaus committed:
fix: Use initializer lists to fix `-Wclazy-container-inside-loop` warnings

Fixes warnings like these:

    src/library/rekordbox/rekordboxfeature.cpp:225:28: error: container inside loop causes unneeded allocations [-Wclazy-container-inside-loop]
                QList<QString> data;
                               ^

Holzhaus committed:
chore(seratofeature): Replace `foreach` with range loop

Holzhaus committed:
ci(github): Enable some more clazy checks from the "manual" level

Holzhaus commented:
I removed the noisy `returning-void-expression` and `missing-qobject-cast` checks for now. This can happen later. But I found some useful checks from the "manual" level that make sense to enable.

Holzhaus committed:
fix(multilineeditdelegate): Fix silenced `-Wclazy-base-class-event`

Swiftb0y commented:
[`container-anti-pattern`](https://github.com/KDE/clazy/blob/master/docs/checks/README-container-anti-pattern.md#container-anti-pattern) is also worth to consider in a follow up PR.

Swiftb0y reviewed (approved):
lgtm thank you. waiting for @daschuer

daschuer reviewed (changes requested):
Just one remaining comment. 
Thank you.

Holzhaus committed:
chore(effectsbackend): Add translator comments for backend type names

Holzhaus commented on src/effects/backends/effectsbackend.cpp:
```diff
@@ -29,11 +29,15 @@ QString EffectsBackend::backendTypeToString(EffectBackendType backendType) {
 QString EffectsBackend::translatedBackendName(EffectBackendType backendType) {
     switch (backendType) {
     case EffectBackendType::BuiltIn:
-        //: Used for effects that are built into Mixxx
```
Done.

daschuer reviewed (approved):
LGTM, Thank you.

Holzhaus commented:
Thanks!
